File: tools/sdks/swiftui/Sources/ThunderSwiftUI/components/presentation/auth/AcceptInvite.swift
import SwiftUI
import Thunder

/// Drives the invited-user-registration flow for a given invitation code (spec §8.4 Presentation).
public struct ThunderAcceptInvite: View {
    @EnvironmentObject private var i18n: ThunderI18n
    public let invitationCode: String
    public let applicationId: String
    public let onComplete: (() -> Void)?
    public let onError: ((String) -> Void)?

    public init(
        invitationCode: String,
        applicationId: String,
        onComplete: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil
    ) {
        self.invitationCode = invitationCode
        self.applicationId = applicationId
        self.onComplete = onComplete
        self.onError = onError
    }

    public var body: some View {
        BaseThunderAcceptInvite(
            invitationCode: invitationCode,
            applicationId: applicationId,
            onComplete: onComplete,
            onError: onError
        ) { signUpState in
            VStack(alignment: .leading, spacing: 12) {
                Text(i18n.resolve("acceptInvite.title"))
                    .accessibilityAddTraits(.isHeader)
                if let error = signUpState.error {
                    Text(error).foregroundColor(.red)
                }
                ForEach(signUpState.inputs, id: \.name) { input in
                    if input.type == "password" {
                        SecureField(input.name, text: signUpState.binding(for: input.name))
                            .frame(minHeight: 44)
                    } else {
                        TextField(input.name, text: signUpState.binding(for: input.name))
                            .frame(minHeight: 44)
                    }
                }
                ForEach(signUpState.actions, id: \.id) { action in
                    Button(action.label ?? i18n.resolve("acceptInvite.submit")) {
                        signUpState.submit(actionId: action.id)
                    }
                    .disabled(signUpState.isLoading)
                    .frame(minWidth: 44, minHeight: 44)
                }
            }
            .padding()
        }
    }
}

/// Unstyled base variant (spec §8.3). Reuses ThunderSignUpState for field/action management.
public struct BaseThunderAcceptInvite<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let invitationCode: String
    public let applicationId: String
    public let onComplete: (() -> Void)?
    public let onError: ((String) -> Void)?
    public let content: (ThunderSignUpState) -> Content

    @StateObject private var signUpState = ThunderSignUpState(submit: { _, _, _ in })

    public init(
        invitationCode: String,
        applicationId: String,
        onComplete: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil,
        @ViewBuilder content: @escaping (ThunderSignUpState) -> Content
    ) {
        self.invitationCode = invitationCode
        self.applicationId = applicationId
        self.onComplete = onComplete
        self.onError = onError
        self.content = content
    }

    public var body: some View {
        content(signUpState).task { await initFlow() }
    }

    private func initFlow() async {
        signUpState.isLoading = true
        defer { signUpState.isLoading = false }
        do {
            let payload = EmbeddedSignInPayload(
                actionId: "__initiate__",
                inputs: ["invitationCode": invitationCode]
            )
            let response = try await state.client.signUp(payload: payload)
            await handleResponse(response)
        } catch {
            signUpState.error = error.localizedDescription
            onError?(error.localizedDescription)
        }
    }

    private func handleResponse(_ response: EmbeddedFlowResponse) async {
        switch response.flowStatus {
        case .complete:
            await state.refresh()
            onComplete?()
        case .promptOnly:
            signUpState.update(from: response)
        case .error:
            let msg = response.failureReason ?? "Could not accept invite"
            signUpState.error = msg
            onError?(msg)
        }
    }
}

File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/auth/AcceptInvite.kt
package io.thunder.compose.components.presentation.auth

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.thunder.android.*
import io.thunder.compose.LocalThunder
import io.thunder.compose.ThunderState
import io.thunder.compose.components.actions.BaseThunderSignUpButton
import kotlinx.coroutines.launch

/** Drives the invited-user-registration flow for a given invitation code (spec §8.4 Presentation). */
@Composable
fun ThunderAcceptInvite(
    invitationCode: String,
    modifier: Modifier = Modifier,
    onComplete: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
) {
    val thunderState = LocalThunder.current
    val i18n = thunderState.i18n
    BaseThunderAcceptInvite(invitationCode = invitationCode, modifier = modifier, onComplete = onComplete, onError = onError) { state ->
        Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            BasicText(i18n.resolve("acceptInvite.title"))
            state.error?.let { BasicText(it) }
            state.inputs.forEach { input ->
                BasicTextField(
                    value = state.fieldValue(input.name),
                    onValueChange = { state.setField(input.name, it) },
                    modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp),
                )
            }
            state.actions.forEach { action ->
                BaseThunderSignUpButton(label = action.label ?: i18n.resolve("acceptInvite.submit")) {
                    state.submit(action.id)
                }
            }
        }
    }
}

/** Unstyled base variant (spec §8.3). Reuses [ThunderSignUpState]. */
@Composable
fun BaseThunderAcceptInvite(
    invitationCode: String,
    modifier: Modifier = Modifier,
    onComplete: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
    content: @Composable (ThunderSignUpState) -> Unit,
) {
    val thunderState = LocalThunder.current
    val scope = rememberCoroutineScope()
    val state = remember { ThunderSignUpState() }

    state.onSubmit = { actionId ->
        scope.launch {
            state.isLoading = true; state.error = null
            try {
                val payload = EmbeddedSignInPayload(flowId = state.flowId, actionId = actionId, inputs = state.fields())
                val response = thunderState.client.signUp(payload = payload)
                handleAcceptResponse(response, state, thunderState, onComplete, onError)
            } catch (e: Exception) {
                state.error = e.message
                onError?.invoke(e.message ?: "Could not accept invite")
            } finally { state.isLoading = false }
        }
    }

    LaunchedEffect(invitationCode) {
        state.isLoading = true
        try {
            val payload = EmbeddedSignInPayload(actionId = "__initiate__", inputs = mapOf("invitationCode" to invitationCode))
            val response = thunderState.client.signUp(payload = payload)
            handleAcceptResponse(response, state, thunderState, onComplete, onError)
        } catch (e: Exception) {
            state.error = e.message
            onError?.invoke(e.message ?: "Could not accept invite")
        } finally { state.isLoading = false }
    }

    Box(modifier = modifier) { content(state) }
}

private suspend fun handleAcceptResponse(
    response: EmbeddedFlowResponse,
    state: ThunderSignUpState,
    thunderState: ThunderState,
    onComplete: (() -> Unit)?,
    onError: ((String) -> Unit)?,
) {
    when (response.flowStatus) {
        FlowStatus.COMPLETE -> { thunderState.refresh(); onComplete?.invoke() }
        FlowStatus.PROMPT_ONLY -> state.update(response)
        FlowStatus.ERROR -> { state.error = response.failureReason; onError?.invoke(response.failureReason ?: "Failed") }
    }
}
